### PR TITLE
docs(core): clarify InjectionToken usage with non-reified types

### PR DESCRIPTION
**Pull request by luishcastroc**
- Emphasized the importance of using the same InjectionToken instance for both provider and injection call.
- Added examples to illustrate correct and incorrect usages to prevent NullInjectorError.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

The current documentation for InjectionToken provides an example where a new instance of InjectionToken is passed to injector.get(). This can be misleading, especially when dealing with types that do not have a runtime representation, like interfaces. The documentation does not explicitly mention the potential issue of receiving a NullInjectorError if different instances of InjectionToken are used in the provider and the injection call.

Issue Number: #51270 

## What is the new behavior?

Documentation updated adding an example on proper usage and explanation.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/di/injection_token.ts [hunk 1]
```diff
@@ -21,10 +21,25 @@ import {ɵɵdefineInjectable} from './interface/defs';
  * `InjectionToken` is parameterized on `T` which is the type of object which will be returned by
  * the `Injector`. This provides an additional level of type safety.
  *
+ * > **Important Note**: When using `InjectionToken` with types that do not have a runtime
```
Comment (JoostK) [inline packages/core/src/di/injection_token.ts:24]: The "not have a runtime representation" is a red-herring here, that aspect is irrelevant in terms of how injection tokens always require referential identity between providers and consumers.

File: packages/core/src/di/injection_token.ts [hunk 2]
```diff
@@ -21,10 +21,25 @@ import {ɵɵdefineInjectable} from './interface/defs';
...
+ *
+ * ```typescript
+ * interface MyInterface {...}
+ * // Correct usage:
+ * const TOKEN = new InjectionToken<MyInterface>('SomeToken');
+ * const myInterface = injector.get(TOKEN);
+ * // myInterface is inferred to be MyInterface.
  * ```
+ *
+ * **Incorrect usage** (leading to `NullInjectorError`):
```
Comment (JoostK) [inline packages/core/src/di/injection_token.ts:38]: I wonder if this shouldn't just be deleted.

File: packages/core/src/di/injection_token.ts [hunk 3]
```diff
@@ -21,10 +21,25 @@ import {ɵɵdefineInjectable} from './interface/defs';
...
+ * representation (like interfaces), ensure that you use the same instance of the `InjectionToken`
+ * in both the provider and the injection call. Creating a new instance of `InjectionToken` in
+ * different places, even with the same description, will be treated as different tokens by
+ * Angular's DI system, leading to a `NullInjectorError`.
+ *
+ * ```typescript
+ * interface MyInterface {...}
+ * // Correct usage:
+ * const TOKEN = new InjectionToken<MyInterface>('SomeToken');
+ * const myInterface = injector.get(TOKEN);
```
Comment (JoostK) [inline packages/core/src/di/injection_token.ts:34]: I'd propose that creating `injector` is actually included in this example, as that would show how the same token instance is used to setup the provider.

File: packages/core/src/di/injection_token.ts [hunk 4]
```diff
@@ -21,10 +21,25 @@ import {ɵɵdefineInjectable} from './interface/defs';
  * `InjectionToken` is parameterized on `T` which is the type of object which will be returned by
  * the `Injector`. This provides an additional level of type safety.
  *
+ * > **Important Note**: When using `InjectionToken` with types that do not have a runtime
+ * representation (like interfaces), ensure that you use the same instance of the `InjectionToken`
+ * in both the provider and the injection call. Creating a new instance of `InjectionToken` in
+ * different places, even with the same description, will be treated as different tokens by
+ * Angular's DI system, leading to a `NullInjectorError`.
+ *
+ * ```typescript
```
Comment (JoostK) [inline packages/core/src/di/injection_token.ts:30]: You may want to use a docregion in an aio test, those are actually executed to ensure they compile.

File: packages/core/src/di/injection_token.ts (hunk 2 above)
Comment (luishcastroc) [reply packages/core/src/di/injection_token.ts:38]: sounds fair, documentation needs to point the proper way and we assume others might be wrong unless pointed differently, i'll make the change.

File: packages/core/src/di/injection_token.ts (hunk 3 above)
Comment (luishcastroc) [reply packages/core/src/di/injection_token.ts:34]: can you please be more specific about the suggestion not sure if i'm following

File: packages/core/src/di/injection_token.ts (hunk 4 above)
Comment (luishcastroc) [reply packages/core/src/di/injection_token.ts:30]: thanks but not really following, can you please point me to an example of usage?

File: packages/core/src/di/injection_token.ts (hunk 1 above)
Comment (luishcastroc) [reply packages/core/src/di/injection_token.ts:24]: what about like this:

```bazel
* <div class="alert is-helpful">
* 
* **Important Note**: Ensure that you use the same instance of the `InjectionToken` in both the provider 
* and the injection call. Creating a new instance of `InjectionToken` in different places, even with the same 
* description, will be treated as different tokens by Angular's DI system, leading to a `NullInjectorError`.
*
* ```typescript
* interface MyInterface {...}
* const TOKEN = new InjectionToken<MyInterface>('SomeToken');
* const myInterface = injector.get(TOKEN);
* // myInterface is inferred to be MyInterface.
* ```
 ```

File: packages/core/src/di/injection_token.ts (hunk 4 above)
Comment (JeanMeche) [reply packages/core/src/di/injection_token.ts:30]: I gave some guidance on the discord 👍

File: packages/core/src/di/injection_token.ts (hunk 3 above)
Comment (JoostK) [reply packages/core/src/di/injection_token.ts:34]: Something along the lines of

```ts
const instance: MyInterface = {};
const injector = Injector.create({ providers: [{ provide: TOKEN, useValue: instance }] });
```

File: packages/core/src/di/injection_token.ts (hunk 1 above)
Comment (JoostK) [reply packages/core/src/di/injection_token.ts:24]: I like it :+1:

**Review by jessicajaniuk — APPROVED** [review]
LGTM

**Review by josephperrott — APPROVED** [review]
LGTM

**Comment by jessicajaniuk** [comment]
This PR was merged into the repository by commit 6b5c812b6b4117fd3544aec078bec98afab647c1.

**Pull request closed without merging**